[PIO] Generate filename at build, not in post processing

TD-er opened the pull request:
Generate filename, included in the build
Use post Python scripts to move files to the right directory
Use post Python scripts to generate .bin.gz files.


Still to do:
- [x] Set `PROGNAME` in the PIO env. with the generated file name (so only need to 'compute' file name once)
- [x] Update to Python 3.8
- [x] Edit `before_deploy` script
- [x] Edit Travis script/build
- [x] Check Vagrant script/build
- [x] Check Travis script/build
- [ ] Add work-around to describe pull request builds in the Git description (currently shows only `HEAD`)
- [x] Improve the move/gzip scripts   (.gz not yet supported for ESP32 for example)
- [x] Move file size check to Python code (needs to be a separate check for GZip files)
- [ ] Create a new 2-step OTA file supporting .bin.gz files
- [x] Document

TD-er committed:
[PIO] Generate filename at build, not in post processing

Generate filename, included in the build
Use post Python scripts to move files to the right directory
Use post Python scripts to generate .bin.gz files.

TD-er commented:
@uzi18 This one is special for you :)
This way we can generate all we need at build time, so no need to patch the binary later.
As the bin file is CRC-checked at OTA, it is also no longer needed to include our own CRC.
So we have to think what to do with the CRC object we had.
Is it still needed?

Still have to do a number of things, so it is not yet ready to merge, but please have a look at it.

TD-er commented:
Not sure what I messed up, but a last test build on my node here does not show the filename.
Maybe some last moment cleanup I did before commit and push.
Have to check tomorrow to see what I broke....


Edit:
That was an easy fix... When showing the file name, make sure to actually print it :)

TD-er committed:
[PIO] Fix showing binary file name in sysinfo page

TD-er commented:
The Travis build is failing.
Apparently on Linux we need 4 layers of double quotes, while on Windows it is only a 3 layers.
Will have a look at it tomorrow....

TD-er committed:
[PIO] Attempt to fix missing quote level in Linux

uzi18 commented:
will look at it, sorry did not found time for it

crc/md5 can be ommited/removed

TD-er commented:
> crc/md5 can be ommited/removed

It is also used in the settings file (at least in the security settings right now), that's why I still have left the struct in.

TD-er committed:
[PIO] Fix Windows/Linux detection in number of quote levels

TD-er committed:
[PIO] Final fix for proper quote level per OS + store build platform

TD-er commented:
OK, fixed it for building in Windows and Linux.

```Python
# needed to wrap in a number of double quotes.
# one level for adding it to the list of defines
# another level to have the string quoted in the .cpp file
# somewhere along the line, another level is removed.
def wrap_quotes(str_value):
    if platform.system() == 'Windows':
        return "\"\"\"{}\"\"\"".format(str_value)
    else:
        return '\'\"\"\"{}\"\"\"\''.format(str_value)
```

Also included the build platform as it may be handy for detecting specific issues related to alignment and such.

Build | Windows
-- | --
Build:⋄ | 20106 - Mega
System Libraries:⋄ | ESP82xx Core 9c56ed1f, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2 PUYA support
Git Build:⋄ |  
Plugin Count:⋄ | 14 [Normal]
Build Origin: | Self built!
Build Time:⋄ | May 3 2020 11:44:11
Binary Filename:⋄ | ESP_Easy_mega_20200503_custom_ESP8266_4M1M
Build Platform:⋄ | Windows-10-10.0.18362-SP0


Build | Linux
-- | --
Build:⋄ | 20106 - Mega
System Libraries:⋄ | ESP82xx Core 9c56ed1f, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2 PUYA support
Git Build:⋄ |  
Plugin Count:⋄ | 14 [Normal]
Build Origin: | Self built!
Build Time:⋄ | May 3 2020 11:51:53
Binary Filename:⋄ | ESP_Easy_mega_20200503_custom_ESP8266_4M1M
Build Platform:⋄ | Linux-4.4.0-18362-Microsoft-x86_64-with-Ubuntu-18.04-bionic

TD-er committed:
[PIO] Add Git branch description to build info

TD-er committed:
[PIO] Move build environment to Python 3.8 as 2.7 is no longer supported

TD-er committed:
[PIO] Python 3.8 requirements.txt list

The pip freeze I used on my Windows PC to build

TD-er committed:
[PIO] Updated packages in Python 3.8 requirements.txt list

Updated all packages we use to their latest version.

At least the latest Sphinx does seem to squish the documentation a bit, so we may need to update the template of our documentation.

TD-er committed:
[PIO] Relax PlatformIO version in requirements.txt to >= 4.3.3

TD-er committed:
[PIO] Use Python 3.8 on Vagrant/Ubuntu builds

TD-er committed:
[PIO] GZip and copy files to build_output directory

TD-er committed:
[PIO] Include PlatformIO env dump in build output for diagnostics

TD-er committed:
[PIO] Update docs/requirements.txt

TD-er committed:
[PIO] Compare build sizes when copying the build files.

TD-er committed:
[PIO] Adopt deploy scripts to new situation where Python scripts do all

TD-er committed:
[PIO] Split zip files in before_deploy

Split into a standard one containing the build, one for docs and one for .elf files and rejected files.

TD-er committed:
[PIO] Zip docs dir recursive

uzi18 commented:
nice job, you got a positive flow now ;)
in fact it is a lot easier to manage all build dependent stuff in python/scons

we got something like this:
https://github.com/letscontrolit/ESPEasy/blob/9538a7198aa07a64ce53bf6e35a1d7bd35dea817/platformio_esp32_envs.ini#L20-L24

and it unroll in env dump : 
```
  'BUILD_FLAGS': [ '-DMQTT_MAX_PACKET_SIZE=1024',
                   '-D BUILD_GIT=\'""\'',
                   '-DCONFIG_FREERTOS_ASSERT_DISABLE',
                   '-DCONFIG_LWIP_ESP_GRATUITOUS_ARP',
                   '-DCONFIG_LWIP_GARP_TMR_INTERVAL=30  '
                   '-DPLUGIN_SET_TEST_ESP32'],
```
so  i guess it looks like only need to return this on both platforms: 
`return '\'"{}"\''.format(str_value)`

TD-er commented:
I'm not sure I get where you're heading at.
The main problem we have with string like defines is to get the double quotes to appear in the code after the pre-processor is done.

```c++
#define FOO "Bar"
String foo = F(FOO);
```

For this you really need the define to have the double quotes, as it is impossible to correct this:

```c++
#define FOO Bar
String foo = F("FOO");
```

Believe me, I tried :)

This was the closest I could come up with, but it for sure will not work in F() macros, only when calling a function and I'm not even sure it is working on all compilers.

```c++
#define FOO Bar
#define FOO_TMP (#FOO)
String foo = F(FOO_TMP);
```

TD-er committed:
[PIO] Add missing parts in the included source dir in the build ZIP file

TD-er committed:
[Vagrant] Fix issue where update fails due to Grub interactive update

TD-er committed:
[Vagrant] Show Vagrant build as build origin on sysinfo page

uzi18 commented:
@TD-er ok so maybe it is something related to command line handling as it goes by shell

BTW ... have preparing to next long waited double feature for esp8266/esp32 builds - but you need to wait for PR to late evening when return to home :)

TD-er commented:
Should I merge this first then?

TD-er committed:
[Shelly PLUG-S] Add template + build definition for Shelly PLUG-S

As mentioned a few times on the forum.
[one topic](https://www.letscontrolit.com/forum/viewtopic.php?f=5&t=7699)

uzi18 commented:
not related to build process but webserver/settings

TD-er committed:
[PIO] Upgrade to platformio / platform-espressif8266 @2.5.0

TD-er committed:
[Docs] Describe files in nightly builds
